java-checks/src/main/java/org/sonar/java/checks/spring/SpringScanDefaultPackageCheck.java

```diff
@@ -0,0 +1,171 @@
+import org.sonar.plugins.java.api.tree.NewArrayTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S4602")
+public class SpringScanDefaultPackageCheck extends IssuableSubscriptionVisitor {
+
+  private static final String DEFAULT_ATTRIBUTE = "value";
+
+  private static final Map<String, Set<String>> SCAN_PACKAGE_ATTRIBUTES = buildScanPackageAttributes();
+
+  private static Map<String, Set<String>> buildScanPackageAttributes() {
+    Map<String, Set<String>> map = new HashMap<>();
+
+    map.put("org.springframework.context.annotation.ComponentScan",
+      new HashSet<>(Arrays.asList(DEFAULT_ATTRIBUTE, "basePackages", "basePackageClasses")));
+
+    map.put("org.springframework.boot.autoconfigure.SpringBootApplication",
+      new HashSet<>(Arrays.asList("scanBasePackages", "scanBasePackageClasses")));
+
+    map.put("org.springframework.boot.web.servlet.ServletComponentScan",
+      new HashSet<>(Arrays.asList(DEFAULT_ATTRIBUTE, "basePackages", "basePackageClasses")));
+
+    return map;
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.ANNOTATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    AnnotationTree annotation = (AnnotationTree) tree;
+    Set<String> scanPackageAttributeNames = SCAN_PACKAGE_ATTRIBUTES.get(annotation.symbolType().fullyQualifiedName());
+    if (scanPackageAttributeNames != null) {
+      List<ExpressionTree> scanPackageAttributes = annotation.arguments().stream()
+        .filter(argument -> scanPackageAttributeNames.contains(attributeName(argument)))
+        .collect(Collectors.toList());
+
+      if (scanPackageAttributes.isEmpty()) {
+        checkIfAnnotationIsInDefaultPackage(annotation);
+      } else {
+        checkAnnotationPackageAttributes(annotation, scanPackageAttributes);
+      }
+    }
+  }
+
+  private void checkIfAnnotationIsInDefaultPackage(AnnotationTree annotation) {
+    if (isNodeInDefaultPackage(annotation)) {
+      reportIssue(annotation.annotationType(), "Remove the annotation \"@" + annotation.symbolType().name() +
+        "\" or move the annotated class out of the default package.");
+    }
+  }
+
+  private void checkAnnotationPackageAttributes(AnnotationTree annotation, List<ExpressionTree> scanPackageAttributes) {
+    List<ExpressionTree> values = scanPackageAttributes.stream()
+      .flatMap(SpringScanDefaultPackageCheck::extractValues)
+      .collect(Collectors.toList());
+
+    values.stream()
+      .flatMap(SpringScanDefaultPackageCheck::findEmptyString)
+      .forEach(expression -> reportIssue(expression, "Define packages to scan. Don't rely on the default package."));
+
+    values.stream()
+      .flatMap(SpringScanDefaultPackageCheck::findClassInDefaultPackage)
+      .forEach(identifier -> reportIssue(identifier, "Remove the annotation \"@" + annotation.symbolType().name() +
+        "\" or move the \"" + identifier.name() + "\" class out of the default package."));
```

Comment:
With this approach of reusing the list you can end up with a case where two issues are raised : I don't see this testcase. 
I am not sure this is actually relevant : if you have one attribute that could raise an issue and the other not, should we raise an issue in this case ?

Reply:
Issues are raised on each invalid values of "@ComponentScan", it's intentional. I added a test case to highlight those cases: several empty strings "" and several classes in the default package.